driver/config/config.cpp

```diff
@@ -97,3 +99,270 @@ void writeDSNinfo(const ConnInfo * ci) {
+            key.pop_back();
+        }
+
+        if (key.empty())
+            throw std::runtime_error("key name is missing");
+
+        if (!equal_met)
+            throw std::runtime_error("'=' is missing");
+
+        return key;
+    };
+
+    const auto extract_key_value = [&consume_space, &consume_semi_colons] (std::string & str) {
+        std::string value;
+
+        while (consume_space(str));
+
+        bool stop_at_closing_brace = false;
+        bool stopped_at_closing_brace = false;
+
+        if (!str.empty() && str.front() == '{') {
+            stop_at_closing_brace = true;
+            str.erase(0, 1);
+        }
+
+        while (!str.empty()) {
+            if (stop_at_closing_brace && str.front() == '}') {
+                stopped_at_closing_brace = true;
+                str.erase(0, 1);
+                break;
+            }
+            else if (!stop_at_closing_brace && str.front() == ';') {
+                stopped_at_closing_brace = false;
+                break;
+            }
+            else {
+                value += str.front();
+                str.erase(0, 1);
+            }
+        }
+
+        if (stop_at_closing_brace) {
+            if (!stopped_at_closing_brace)
+                throw std::runtime_error("'}' expected");
+        }
+        else {
+            while (!value.empty() && std::isspace(value.back())) {
+                value.pop_back();
+            }
+        }
+
+        std::size_t consumed = 0;
+        std::size_t consumed_semi_colons = 0;
+
+        do {
+            consumed = consume_semi_colons(str);
+            consumed_semi_colons += consumed;
+            consumed += consume_space(str);
+        } while (consumed > 0);
+
+        if (consumed_semi_colons == 0 && !str.empty())
+            throw std::runtime_error("';' expected");
+
+        return value;
+    };
+
+    while (!cs.empty()) {
+        const auto key = extract_key_name(cs);
+        const auto value = extract_key_value(cs);
+
+        const auto key_upper = Poco::UTF8::toUpper(key);
+
+        if (
+            fields.find(key) != fields.end() ||
+            (key_upper == INI_DSN && fields.find(INI_FILEDSN) != fields.end()) ||
+            (key_upper == INI_FILEDSN && fields.find(INI_DSN) != fields.end()) ||
+            (key_upper == INI_DSN && fields.find(INI_DRIVER) != fields.end()) ||
+            (key_upper == INI_DRIVER && fields.find(INI_DSN) != fields.end())
+        ) {
+            break; // These doesn't override each other: the first one met is used.
+        }
+
+        fields[key] = value;
+    }
+
+    return fields;
+}
+
+key_value_map_t readDSNInfo(const std::string & dsn_utf8) {
+    key_value_map_t fields;
+
+    std::basic_string<CharTypeLPCTSTR> dsn;
+    std::basic_string<CharTypeLPCTSTR> config_file;
+    const std::basic_string<CharTypeLPCTSTR> default_value = fromUTF8<CharTypeLPCTSTR>("__default__"); // Assuming, __default__ will never be used as a value for any key.
```

Comment:
Just curious, what are the chances of having a key named `__default__` ? and can we have a default key to be empty string `std::basic_string{}` ?

Reply:
Since a key can be empty but existing in the DSN, and I needed a way of telling if it is missing or just empty, I had to use some non-empty default value...